### PR TITLE
pass PGPASSWORD via env directly, not via shell

### PR DESCRIPTION
**Pull request by evgeni**

### CODE REVIEW

**Comment by evgeni** [comment]
with this, nothing uses `hidden_patterns` anymore. should we drop it, so that this bad design doesn't creep up on us again?

**Review by ehelms — APPROVED** [review]
Agreed on cleaning up hidden_patterns if it's not used.

File: lib/foreman_maintain/concerns/base_database.rb [hunk 1]
```diff
@@ -146,17 +149,20 @@ def raise_psql_missing_error
       private
 
       def base_command(config, command = 'psql')
-        "PGPASSWORD='#{config[%(password)]}' "\
-        "#{command} -h #{config['host'] || 'localhost'} "\
+        env = { 'PGPASSWORD' => config['password'] }
+        cmd = "#{command} -h #{config['host'] || 'localhost'} "\
         " -p #{config['port'] || '5432'} -U #{config['username']}"
+        return cmd, env
```
[the remark this answers is not in the document]
Comment (evgeni) [reply lib/foreman_maintain/concerns/base_database.rb:155]: and yet it didn't complain :D
(I think it only complains on singular returns, not lists)

File: test/lib/utils/command_runner_test.rb
[the remark this answers is not in the document]
Comment (evgeni) [reply test/lib/utils/command_runner_test.rb:1]: we could, yeah. I pondered writing actual new tests, but didn't so far.

**Comment by ekohl** [comment]
I wonder if it could wrap django's dbshell command to run queries somehow

**Comment by evgeni** [comment]
Don't give me ideas

**Comment by evgeni** [comment]
```python
PYTHONDONTWRITEBYTECODE=1 PYTHONPATH=/etc/pulp python -c 'import yaml; import settings; print(yaml.safe_dump(settings.DATABASES["default"]))'
```

**Comment by ekohl** [comment]
My biggest issue is that the Rails equivalent needs to load the entire Rails env, which is slow. And there is none (that I know of) for Candlepin. Not that we need queries on that after we merge https://github.com/theforeman/foreman_maintain/pull/940, but there's still the backup/restore part.

**Comment by ekohl** [comment]
> PYTHONDONTWRITEBYTECODE=1 PYTHONPATH=/etc/pulp python -c 'import yaml; import settings; print(yaml.safe_dump(settings.DATABASES["default"]))'

Nit: I'd prefer JSON and you can use `pulpcore-admin shell` to get a Python shell with guaranteed the correct Python version.

**Comment by evgeni** [comment]
sure, why not:
```console
# PULP_SETTINGS=/etc/pulp/settings.py pulpcore-manager shell -c 'from django.conf import settings; import json; print(json.dumps(settings.DATABASES["default"]))'
{"ENGINE": "django.db.backends.postgresql", "NAME": "pulpcore", "USER": "pulp", "PASSWORD": "sdfsfds\"dssdfdsf", "HOST": "remotedb", "PORT": "5432", "ATOMIC_REQUESTS": false, "AUTOCOMMIT": true, "CONN_MAX_AGE": 0, "CONN_HEALTH_CHECKS": false, "OPTIONS": {}, "TIME_ZONE": null, "TEST": {"CHARSET": null, "COLLATION": null, "MIGRATE": true, "MIRROR": null, "NAME": null}}
```

**Comment by ekohl** [comment]
The reason I bring it up is that it's probably easier to reuse this once we start using containers. If we start using env vars or secrets, that'll work at runtime but it'll be complex to parse all the options.

**Comment by evgeni** [comment]
https://github.com/theforeman/foreman_maintain/pull/941

File: lib/foreman_maintain/concerns/base_database.rb [hunk 2]
```diff
@@ -125,8 +127,9 @@ def dropdb(config = configuration)
       def db_version(config = configuration)
         if ping(config)
           # Note - t removes headers, -A removes alignment whitespace
-          server_version_cmd = psql_command(config) + ' -c "SHOW server_version" -t -A'
-          version_string = execute!(server_version_cmd, :hidden_patterns => [config['password']])
+          cmd, env = psql_command(config)
+          cmd += ' -c "SHOW server_version" -t -A'
+          version_string = execute!(cmd, :env => env)
```
[the remark this answers is not in the document]
Comment (evgeni) [reply lib/foreman_maintain/concerns/base_database.rb:132]: maybe, but I also didn't feel like *actually* touching the code here

**Review by ehelms — APPROVED** [review]

**Comment by ekohl** [comment]
I wanted to know how I'd structure it and in https://github.com/theforeman/foreman_maintain/pull/945 there are the results of my refactor.

The biggest change is that instead of composing complex commands I pass all PG connection params via env vars which simplifies a lot of the code.

I can add your tests if you'd like (and adopt where needed).

**Comment by evgeni** [comment]
> I can add your tests if you'd like (and adopt where needed).

I think that'd be a good idea, but maybe just merge this and then use yours as further improvement?

**Comment by ekohl** [comment]
> I can add your tests if you'd like (and adopt where needed).

This is done now.

> maybe just merge this and then use yours as further improvement?

I'm not sure I like refactoring it twice shortly after each other.

**Comment by evgeni** [comment]
Strictly speaking this one was never intended as refactoring, just fixing the bug we have when the password contains quotes ;)

**Comment by evgeni** [comment]
closing in favor of #945

**Pull request closed without merging**